File: src/armies/eligibility.py
"""Malus computation and spawn gate logic.

Formula for each ledger entry allocated to a general:
    if decays:
        effective = raw_malus * (share/100) * (0.5 ** (days_since / 14))
    else:
        effective = raw_malus * (share/100)

Total effective malus = sum of all per-entry values.
"""

from __future__ import annotations

import logging
import math
from datetime import date, datetime
from pathlib import Path
from typing import Any

import yaml

log = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Tier gate table — mirrors rules/eligibility.yaml
# ---------------------------------------------------------------------------

TIERS: list[dict[str, Any]] = [
    {
        "min": 0,
        "max": 99,
        "name": "Clean",
        "coordinator": "CLEAR",
        "emergency_reserve": "CLEAR",
        "specialist": "CLEAR",
        "validator": "CLEAR",
    },
    {
        "min": 100,
        "max": 199,
        "name": "Warning",
        "coordinator": "BLOCKED",
        "emergency_reserve": "FOUNDER",
        "specialist": "CLEAR",
        "validator": "CLEAR",
    },
    {
        "min": 200,
        "max": 299,
        "name": "Probation",
        "coordinator": "BLOCKED",
        "emergency_reserve": "BLOCKED",
        "specialist": "REVIEW",
        "validator": "CLEAR",
    },
    {
        "min": 300,
        "max": 399,
        "name": "Demotion risk",
        "coordinator": "BLOCKED",
        "emergency_reserve": "BLOCKED",
        "specialist": "ESCALATE",
        "validator": "CLEAR",
    },
    {
        "min": 400,
        "max": math.inf,
        "name": "Suspension",
        "coordinator": "BLOCKED",
        "emergency_reserve": "BLOCKED",
        "specialist": "BLOCKED",
        "validator": "BLOCKED",
    },
]

KNOWN_ROLES = ("coordinator", "emergency_reserve", "specialist", "validator")


def _parse_date(value: Any) -> date:
    """Parse a date value that may be a string or datetime.date."""
    if isinstance(value, date):
        return value
    if isinstance(value, datetime):
        return value.date()
    # Try ISO string
    return date.fromisoformat(str(value).strip())


def compute_effective_malus(agent_name: str, ledger_path: Path) -> float:
    """Return the total effective malus for *agent_name* from *ledger_path*.

    If the ledger file does not exist, returns 0.0.
    The agent name comparison is case-insensitive.
    """
    if not ledger_path.exists():
        return 0.0

    with ledger_path.open(encoding="utf-8") as fh:
        raw = yaml.safe_load(fh)

    if not isinstance(raw, list):
        return 0.0

    today = date.today()
    total = 0.0
    target = agent_name.strip().lower()

    for entry in raw:
        if not isinstance(entry, dict):
            continue

        raw_malus = float(entry.get("raw_malus", 0))
        decays = bool(entry.get("decays", True))

        # Support both `allocation` list and flat `agent`/`share` fields
        allocations: list[dict[str, Any]] = []

        if "allocation" in entry and isinstance(entry["allocation"], list):
            allocations = entry["allocation"]
        elif "agent" in entry:
            # Flat format: agent + optional share (default 100)
            allocations = [
                {
                    "agent": entry["agent"],
                    "share": entry.get("share", 100),
                }
            ]

        for alloc in allocations:
            if not isinstance(alloc, dict):
                continue
            alloc_agent = str(alloc.get("agent", "")).strip().lower()
            if alloc_agent != target:
                continue

            # Clamp share to [0, 100] — values outside this range are nonsensical
            # and can produce malus contributions larger than raw_malus (#31).
            share = max(0.0, min(100.0, float(alloc.get("share", 100))))

            if decays:
                entry_date_raw = entry.get("date")
                try:
                    entry_date = _parse_date(entry_date_raw)
                    # Clamp to 0 — future-dated entries decay at full strength, not
                    # negative (which would amplify malus instead of decaying it) (#30).
                    days_since = max(0, (today - entry_date).days)
                except (TypeError, ValueError):
                    # Log the bad entry so operators can find and fix it (#21, #40).
                    log.warning(
                        "Malus entry has unparseable date, treating as today: %r", entry
                    )
                    days_since = 0
                contribution = raw_malus * (share / 100) * (0.5 ** (days_since / 14))
            else:
                contribution = raw_malus * (share / 100)

            total += contribution

    return total


def tier_for_malus(effective_malus: float) -> dict[str, Any]:
    """Return the tier dict from TIERS for the given effective malus value."""
    for tier in TIERS:
        if tier["min"] <= effective_malus <= tier["max"]:
            return tier
    # Fallback — should not happen given the last tier has max=inf
    return TIERS[-1]


def eligibility_status(agent_name: str, ledger_path: Path) -> dict[str, Any]:
    """Compute and return full eligibility info for an agent.

    Returns a dict with keys:
        effective_malus  — float
        tier             — tier name string
        roles            — dict[role_name -> status string]
        overall          — "eligible" | "restricted" | "blocked"
    """
    malus = compute_effective_malus(agent_name, ledger_path)
    tier = tier_for_malus(malus)

    roles = {role: tier[role] for role in KNOWN_ROLES}

    # Overall summary
    statuses = set(roles.values())
    if all(s == "CLEAR" for s in statuses):
        overall = "eligible"
    elif "BLOCKED" in statuses and all(
        s == "BLOCKED" for s in statuses
    ):
        overall = "blocked"
    else:
        overall = "restricted"

    return {
        "effective_malus": malus,
        "tier": tier["name"],
        "roles": roles,
        "overall": overall,
    }

File: internal/eligibility/eligibility.go
// Package eligibility computes malus scores, maps them to tiers, and determines
// spawn eligibility for agents. It ports eligibility.py from the Python armies CLI.
package eligibility

import (
	"log/slog"
	"math"
	"time"
)

// nowFunc is the clock source used by EffectiveMalus. Override via SetNowFunc in tests.
var nowFunc = func() time.Time { return time.Now() }

// SetNowFunc replaces the clock function and returns the previous one.
// Use in tests to pin time; always defer restoring the original.
func SetNowFunc(f func() time.Time) func() time.Time {
	old := nowFunc
	nowFunc = f
	return old
}

// KnownRoles returns every role the eligibility system tracks.
// A function is used rather than a variable to prevent callers from mutating the slice.
func KnownRoles() []string {
	return []string{"coordinator", "emergency_reserve", "specialist", "validator"}
}

// Tier describes a malus band and the spawn gate for each role within that band.
type Tier struct {
	Name             string
	Min              float64
	Coordinator      string
	EmergencyReserve string
	Specialist       string
	Validator        string
}

// MalusEntry is one line from an agent's malus ledger.
type MalusEntry struct {
	RawMalus float64
	Share    float64
	Date     string
	Decays   bool
}

// Status is the computed eligibility result for one agent.
type Status struct {
	EffectiveMalus float64
	Tier           string
	Roles          map[string]string
	Overall        string // "eligible" | "restricted" | "blocked"
}

// tiers is the ordered tier table. Entries are half-open intervals [Min, next.Min).
// The last entry covers [400, ∞). Keep in ascending Min order.
var tiers = []Tier{
	{Name: "Clean", Min: 0, Coordinator: "CLEAR", EmergencyReserve: "CLEAR", Specialist: "CLEAR", Validator: "CLEAR"},
	{Name: "Warning", Min: 100, Coordinator: "BLOCKED", EmergencyReserve: "FOUNDER", Specialist: "CLEAR", Validator: "CLEAR"},
	{Name: "Probation", Min: 200, Coordinator: "BLOCKED", EmergencyReserve: "BLOCKED", Specialist: "REVIEW", Validator: "CLEAR"},
	{Name: "Demotion risk", Min: 300, Coordinator: "BLOCKED", EmergencyReserve: "BLOCKED", Specialist: "ESCALATE", Validator: "CLEAR"},
	{Name: "Suspension", Min: 400, Coordinator: "BLOCKED", EmergencyReserve: "BLOCKED", Specialist: "BLOCKED", Validator: "BLOCKED"},
}

// TierFor returns the Tier that contains effectiveMalus.
// The input is rounded to 10 decimal places to eliminate floating-point noise
// before the interval comparison.
func TierFor(effectiveMalus float64) Tier {
	rounded := math.Round(effectiveMalus*1e10) / 1e10

	// Walk backwards: the last tier whose Min ≤ rounded wins.
	matched := tiers[0]
	for _, t := range tiers {
		if rounded >= t.Min {
			matched = t
		}
	}
	return matched
}

// EffectiveMalus computes the total effective malus from a pre-filtered list of entries.
// Caller is responsible for filtering entries to a single agent (case-insensitive) and
// expanding allocation lists into per-entry rows before calling this function.
func EffectiveMalus(entries []MalusEntry) float64 {
	today := nowFunc()
	var total float64

	for _, e := range entries {
		// Clamp share to [0, 100].
		share := max(0.0, min(100.0, e.Share))

		var contribution float64
		if e.Decays {
			t, err := time.Parse("2006-01-02", e.Date)
			if err != nil {
				slog.Warn("malus entry has unparseable date, treating as today", "date", e.Date)
				daysSince := 0.0
				// Half-life of 14 days: factor = 0.5^(days/14)
				decayFactor := math.Pow(0.5, daysSince/14)
				contribution = e.RawMalus * (share / 100) * decayFactor
			} else {
				daysSince := today.Sub(t).Hours() / 24
				if daysSince < 0 {
					// Future date → clamp to 0.
					daysSince = 0
				}
				// Half-life of 14 days: factor = 0.5^(days/14)
				decayFactor := math.Pow(0.5, daysSince/14)
				contribution = e.RawMalus * (share / 100) * decayFactor
			}
		} else {
			contribution = e.RawMalus * (share / 100)
		}

		total += contribution
	}
	return total
}

// EligibilityStatus computes eligibility for an agent.
// agentName is accepted for parity with the Python API and future logging/metrics use.
func EligibilityStatus(agentName string, effectiveMalus float64) Status {
	tier := TierFor(effectiveMalus)

	roles := map[string]string{
		"coordinator":       tier.Coordinator,
		"emergency_reserve": tier.EmergencyReserve,
		"specialist":        tier.Specialist,
		"validator":         tier.Validator,
	}

	// Determine overall gate from the role values.
	allClear := true
	allBlocked := true
	for _, v := range roles {
		if v != "CLEAR" {
			allClear = false
		}
		if v != "BLOCKED" {
			allBlocked = false
		}
	}

	var overall string
	switch {
	case allClear:
		overall = "eligible"
	case allBlocked:
		overall = "blocked"
	default:
		overall = "restricted"
	}

	return Status{
		EffectiveMalus: effectiveMalus,
		Tier:           tier.Name,
		Roles:          roles,
		Overall:        overall,
	}
}
